nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/RunMongoAggregation.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.bson.conversions.Bson;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Tags({"mongo", "aggregation", "aggregate"})
+@CapabilityDescription("A processor that runs an aggregation query whenever a flowfile is received.")
+@InputRequirement(InputRequirement.Requirement.INPUT_ALLOWED)
+@EventDriven
+public class RunMongoAggregation extends AbstractMongoProcessor {
+
+    private final static Set<Relationship> relationships;
+    private final static List<PropertyDescriptor> propertyDescriptors;
+
+    static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .description("The input flowfile gets sent to this relationship when the query succeeds.")
+            .name("original")
+            .build();
+    static final Relationship REL_FAILURE = new Relationship.Builder()
+            .description("The input flowfile gets sent to this relationship when the query fails.")
+            .name("failure")
+            .build();
+    static final Relationship REL_RESULTS = new Relationship.Builder()
+            .description("The result set of the aggregation will be sent to this relationship.")
+            .name("results")
+            .build();
+
+    static final List<Bson> buildAggregationQuery(String query) throws IOException {
+        List<Bson> result = new ArrayList<>();
+
+        ObjectMapper mapper = new ObjectMapper();
+        List<Map> values = mapper.readValue(query, List.class);
+        for (Map val : values) {
+            result.add(new BasicDBObject(val));
+        }
+
+        return result;
+    }
+
+    public static final Validator AGG_VALIDATOR = (subject, value, context) -> {
+        final ValidationResult.Builder builder = new ValidationResult.Builder();
+        builder.subject(subject).input(value);
+
+        if (context.isExpressionLanguageSupported(subject) && context.isExpressionLanguagePresent(value)) {
+            return builder.valid(true).explanation("Contains Expression Language").build();
+        }
+
+        String reason = null;
+        try {
+            buildAggregationQuery(value);
+        } catch (final RuntimeException | IOException e) {
+            reason = e.getLocalizedMessage();
+        }
+
+        return builder.explanation(reason).valid(reason == null).build();
+    };
+
+    static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+            .name("mongo-agg-query")
+            .displayName("Query")
+            .expressionLanguageSupported(true)
+            .description("The aggregation query to be executed.")
```

Comment:
That's a fair point since the description doesn't allow pretty printing. However it would make a good section in an additionalDetails.html for the overall doc, and you could refer to the Additional Details section here in the description.

Reply:
I'll look for an example of that then.

Reply:
Ok, borrowed an example from the Kafka processors.